packages = c("tidyverse","Biobase","limma","GSEABase","sigPathway","readxl","GSVA")

## Now load or install&load all
package.check <- lapply(
        packages,
        FUN = function(x) {
                if (!require(x, character.only = TRUE)) {
                        BiocManager::install(x)
                        library(x, character.only = TRUE)
                }
        }
)

## 
day <- 7
#countfile<-read.delim(file = "p470/Day0/extended/p470.counts.txt",check.names = FALSE,row.names = 1)
#the count file that is read by read.delim depends on the day argument above and
countfile<-read.delim(file = file.choose(new = "Find the countfile of the required day"),check.names = FALSE,row.names = 1)

#the countfile must be read as a matrix to be used in gsva function

eset_matrix <- as.matrix(countfile)

#we import the genesets
btmgeneset<-gmxToG("BTM gene sets.gmx")# from the sigpathway package
library(sigPathway)
porsign<-read.gmt("modules/PorSignDB.gmt")
library(clusterProfiler)
btmgeneset<-read.gmt("~/Documents/Bioinformatics/Projects/PRRSV project/PRRSV/BTM gene sets.gmx")
#small loop to create a genesets list to label them in the dataframe
genesets_list<-list()
genesets_title<-c()
for (i in 1:346){
        
        genesets_title[i]<-btmgeneset[[i]]$title
        genesets_list[[i]]<-btmgeneset[[i]]$probes}

results_gsva<-lapply(X = btmgeneset,FUN = function(x){gsva(expr=eset_matrix, 
                                                           gset.idx.list=x, 
                                                           annotation,
                                                           method="gsva",#we use the gsva method for the Enrichment scores
                                                           kcdf="Poisson",#poisson for discrete numbers
                                                           abs.ranking=FALSE,
                                                           min.sz=1,#minimal size of each genesets
                                                           max.sz=Inf,#maximal size of each genesets
                                                           parallel.sz=0,
                                                           parallel.type="SOCK",
                                                           mx.diff=TRUE,
                                                           tau=1,#tau=switch(method,gsva=1, ssgsea=0.25, NA),
                                                           ssgsea.norm=TRUE,
                                                           verbose=TRUE)})

results_gsva_test<-gsva(expr=eset_matrix, 
                   gset.idx.list=btmgeneset,
                   annotation,
                   method="gsva",#we use the gsva method for the Enrichment scores
                   kcdf="Poisson",#poisson for discrete numbers
                   abs.ranking=FALSE,
                   min.sz=1,#minimal size of each genesets
                   max.sz=Inf,#maximal size of each genesets
                   parallel.sz=3,
                   mx.diff=TRUE,
                   tau=1,#tau=switch(method,gsva=1, ssgsea=0.25, NA),
                   ssgsea.norm=TRUE,
                   verbose=TRUE)
##Loading the database of signatures
pig_gs<-msigdbr::msigdbr(species = "Sus scrofa")
# sfiltering of the category C7 which only deals with immune response
pig_gs_c7<-filter(pig_gs,gs_cat=="C7")
ensembl<-bitr(pig_gs_c7$entrez_gene,"ENTREZID","UNIGENE",OrgDb = "org.Ss.eg.db")
results_gsva_porsign<-gsva(expr=eset_matrix, 
                        gset.idx.list=porsign,
                        annotation,
                        method="gsva",#we use the gsva method for the Enrichment scores
                        kcdf="Poisson",#poisson for discrete numbers
                        abs.ranking=FALSE,
                        min.sz=1,#minimal size of each genesets
                        max.sz=Inf,#maximal size of each genesets
                        parallel.sz=3,
                        mx.diff=TRUE,
                        tau=1,#tau=switch(method,gsva=1, ssgsea=0.25, NA),
                        ssgsea.norm=TRUE,
                        verbose=TRUE)

results_gsva_df<-data.frame(matrix(unlist(results_gsva),#we transform the list in a dataframe
                                   nrow = length(results_gsva),
                                   byrow = T),
                            stringsAsFactors = FALSE)

results_gsva_df<-t(results_gsva_df)#we transpose because we want the genesets as col and samples as rows

sampleid<-colnames(countfile)#we take the names of the samples from the countfile

colnames(results_gsva_df)<- genesets #the genesets names are taken from the genesets vector created earlier

#add the rownames of the samples (now as rows)
results_gsva_df$sampleid<-sampleid

#assign a new name to the data frame with Day + day of the count file + _ES (enrichment score)
assign(paste("Day",day,"_ES",sep = ""),results_gsva_df)

######################################################################################################################################
######################################################################################################################################
###################At this point the 3 data frames  with the enrichment score should be ready#######################
######################################################################################################################################
######################################################################################################################################

#we bind the 3 data frames by rows
completeES<-rbind(Day0_ES_t,Day3_ES_t,Day7_ES_t)

#make a samplid column containing the infos about the  samples taken from the rownames
#completeES$sampleid <- row.names(completeES)


#we do a pivot of the Enrichment score (ES) to get only one column with our ES
completeES_long<-pivot_longer(completeES,
                              cols = -sampleid,
                              names_to = "genesets",
                              values_to = "es")

# we clean the sample id column
completeES_long$sampleid<-gsub("_","",completeES_long$sampleid)# we replace the "_" by ""

## we replace "CH431" by "" in order to use the ID number for merging the data set
completeES_long$sampleid<-gsub("CH431","",completeES_long$sampleid)
samplesinfos$sampleid<-gsub("CH431-","",samplesinfos$sampleid)

## we set the sampleid col to numeric in the ES and the sampleinfos file
completeES_long$sampleid<-as.numeric(completeES_long$sampleid)

#import the data frame containing the info about the samples
samplesinfos <- read_excel("CH431.xlsx", 
                           col_names = FALSE, skip = 2,col_types = c("text"))
#we assign new names for the columns
colnames(samplesinfos)<- c("sampleid",
                           "virus",
                           "day",
                           "code",
                           "animalid",
                           "treatment",
                           "rqn")

samplesinfos$sampleid<-as.numeric(samplesinfos$sampleid)

#######Major step we merge the ES data with the sample informations ####

completeES_long_merge<-merge(completeES_long,samplesinfos)


completeES_long_merge$day<-gsub("Day","",completeES_long_merge$day)

#we remove the trailing whitespace
completeES_long_merge$day<-str_trim(completeES_long_merge$day)

#we transform to factors
completeES_long_merge$genesets<-as.factor(as.character(completeES_long_merge$genesets))
str(completeES_long_merge)

#we take the proliferation data sets and do some cleaning

proliferation <- read_excel("T cell data/Kick_PRRSV_proliferation.xlsx", 
                            skip = 1)
proliferation_longer<-pivot_longer(proliferation,cols=-1:-5)
colnames(proliferation_longer)<-tolower(colnames(proliferation_longer))
proliferation_longer$name<-as.factor(tolower(proliferation_longer$name))
names(proliferation_longer)<-c("well","day","treatment","stimulus","animalid","name","value" )

#we clean certain columns 
completeES_long_merge$treatment<-word(completeES_long_merge$treatment,1)#we extract the treatment value in a simpler form now ("MLV","MOCK"...)
completeES_long_merge$day<-factor(as.character(completeES_long_merge$day))
completeES_long_merge$animalid<-factor(completeES_long_merge$animalid)
#
proliferation_longer$animalid<-factor(proliferation_longer$animalid)
proliferation_longer$day<-factor(as.character(proliferation_longer$day))
##We change the values of the stimulus (1-3-4=LP,1-7-3 =HP,MED=MOCK,MLV= vr2234)
proliferation_longer$stimulus<-as.factor(proliferation_longer$stimulus)
levels(proliferation_longer$stimulus)<- c("LP","HP","MOCK","MLV")
########
## we calculate the adjusted value for each value of proliferation, ie we sub-
## subtract the value of the stimulation with "MED" (MOCK) to the homologous
## stimulation

test_valueadj<-as.data.frame(proliferation_longer) 
test_valueadj$value<-replace_na(test_valueadj$value,replace = 0)
test_test_valueadj_MOCK <- filter(test_valueadj, stimulus == "MOCK")

test_test_valueadj_NONMOCK <- filter(test_valueadj, Group != "MOCK") 

test_valueadj <- left_join(
        test_valueadj,
        test_test_valueadj_MOCK,
        by = c("animal", "Day", "name","Group"),
        suffix = c("", "_mock")
) %>%
        mutate(adjusted = value - value_mock)

test_valueadj<-filter(test_valueadj,Group==stimulus)

test_valueadj$valueadj<-ifelse(sign(test_valueadj$adjusted)==-1,0,test_valueadj$adjusted)

proliferation_longer_adjusted<-test_valueadj[-c(3,7,8,9)]

colnames(proliferation_longer_adjusted)<- c("day","treatment","animalid","name","value","valueadj")
```
-------
#for the last step we merge together the data sets ES and proliferation data in 1 super data frame called superdf

superdf<-merge(proliferation_longer,completeES_long_merge,by=c("animalid","treatment"),all = TRUE)

##create a superdf using the ratio between the days7 and day3 /day0
## we proliferation_clean which contains only the homologous treatment MOCK MOCK HP HP etc
superdf_ratio_clean <- merge(proliferation_longer_clean,completeES_long_merge_ratio,by=c("animalid"),all = TRUE)
superdfadj_abtiters<-merge()
##some graphs to explore the data 
graph2<-superdf%>%
        filter(name=="%prol. t-cells")%>%
        filter(day.x=="28")%>%
        filter(day.y=="3")%>%
        filter(genesets=="M0")%>%
        ggplot(aes(x=es,y=value,shape=stimulus,color=animalid))+
        geom_point()+
        ylab("%prol. t-cells")+
        xlab(label = "enrichment score")+
        theme_classic()+
        stat_smooth(method = "lm",col="black")+
        theme(aspect.ratio = 0.8)+
        facet_wrap(.~treatment)
graph3<-superdf_clean%>%
        filter(name=="%prol. t-cells")%>%
        filter(day.x=="42")%>%
        filter(day.y==c("7"))%>%
        filter(genesets=="M67")%>%
        #cor(x = superdf$es,y = superdf$value,use="everything",method="pearson")%>%
        ggplot(aes(x=es,y=value,shape=stimulus,color=animalid))+
        geom_point()+
        xlab(label = "enrichment score")+
        ylab("%prol. t-cells")+
        stat_smooth(method = "lm",col="black")+
        ggtitle("M67")+
        theme_classic()+
        theme(aspect.ratio = 0.8)+
        facet_wrap(.~treatment)
print(graph3)
graph4<- superdf%>%
        filter(name=="%prol. treg" )%>%
        ggplot(aes(x=day.x,y=value),color=treatment)+
        geom_boxplot()+
        facet_wrap(~treatment)
print(graph4)
graphs<-ggarrange(graph2,graph3,graph4,common.legend = TRUE,labels = list("Day56vsDay7","Day28vsDay7"))
print(graphs)        

